Source/Cypher/app/src/main/java/com/semaphore_soft/apps/cypher/networking/Server.java

```diff
@@ -0,0 +1,172 @@
+    private int                      maxPlayers     = 4;
+
+
+    public Server()
+    {
+    }
+
+    public void startAcceptor()
+    {
+        AcceptorThread acceptorThread = new AcceptorThread();
+        acceptorThread.start();
+    }
+
+    public void writeAll(String str)
+    {
+        Log.d("Server", "Attempting to write to all clients");
+        for (ClientHandler server : clients)
+        {
+            server.write(str);
+        }
+    }
+
+    public void writeToClient(String str, int index)
+    {
+        Log.d("Threads", "Attempting to write to client " + String.valueOf(index));
+        // TODO call writeALL if index invalid?
+        // Service will default to -1 if no index is given
+        if (!clients.isEmpty() && index > 0)
+        {
+            clients.get(index).write(str);
+        }
+    }
+
+    private class AcceptorThread extends Thread
+    {
+        // The local server socket
+        ServerSocket serverSocket = null;
+        Socket mySocket;
+
+        public AcceptorThread()
+        {
+            try
+            {
+                serverSocket = new ServerSocket(MainActivity.SERVER_PORT);
+
+            }
+            catch (IOException e)
+            {
+                e.printStackTrace();
+                Log.e("ClientHandler", "Failed to start server");
+                Toast.makeText(mContext, "Failed to start server", Toast.LENGTH_SHORT).show();
+            }
+        }
+
+        public void run()
+        {
+            accepting = true;
+            int id = 0;
+            while (accepting && id < maxPlayers - 1)
+            {
+                try
+                {
+                    Log.i("ClientHandler", "Waiting on accept");
+                    mServiceIntent.setData(Uri.parse(NetworkConstants.THREAD_UPDATE));
+                    mServiceIntent.putExtra(NetworkConstants.MSG_EXTRA, "Waiting on accept");
+                    mContext.startService(mServiceIntent);
+
+                    mySocket = serverSocket.accept();
+                    ClientHandler serverThread = new ClientHandler(mySocket, id);
+                    clients.add(serverThread);
+                    serverThread.start();
+                    id++;
+                    accepting = false;
+                }
+                catch (SocketException e)
+                {
+                    Log.i("AcceptorThread", "Socket closed");
+                }
+                catch (IOException e)
+                {
+                    e.printStackTrace();
+                }
+            }
+        }
+    }
+
+    private class ClientHandler extends Thread
+    {
+        // The local server socket
+        Socket mySocket;
+        int    id;
+
+        public ClientHandler(Socket socket, int id)
+        {
+            mySocket = socket;
+            this.id = id;
+        }
+
+        public void run()
+        {
+            Boolean running = true;
+            while (running)
+            {
+                try
```

Comment:
I'd consider extracting the contents of this while loop to a read function, just for organization sake.